transformations/contextual_meaning_perturbation/transformation.py

```diff
@@ -0,0 +1,172 @@
+
+    def get_linguistic_features(self, input: str):
+        """
+        Linguistic analysis of the input sentence.
+        Returns a list of tokens, POS tags and lemmatised tokens.
+        """
+        tokens = []
+        pos_tags = []
+        lemmas = []
+        sentence = self.linguistic_pipeline(input)
+        for token in sentence:
+            tokens.append(token.text)
+            pos_tags.append(token.pos_)
+            lemmas.append(token.lemma)
+        return tokens, pos_tags, lemmas
+
+    def count_eligible_tokens(self, pos_tags):
+        """
+        Returns the number of words that will be replaced.
+        """
+        pos_count = 0
+        for pos in self.pos_to_change:
+            pos_count += pos_tags.count(pos)
+        return pos_count
+
+    def get_perturbation_candidates(self, input: str):
+        """
+        Samples tokens that will be replaced.
+        Generates and returns the top k contextual replacement candidates.
+        """
+        perturbation_dict = {}
+        tokens, pos_tags, lemmas = self.get_linguistic_features(input)
+        if self.count_eligible_tokens(pos_tags) == 0:
+            print(
+                "Warning: The sequence' ",
+                input,
+                " 'will remain unchanged as it didn't include the following POS tags:",
+                self.pos_to_change,
+            )
+            num_word_swaps = 0
+        else:
+            num_word_swaps = max(
+                1,
+                round(self.percentage * self.count_eligible_tokens(pos_tags)),
+            )
+        if self.verbose:
+            print(num_word_swaps, " tokens will be masked.\n")
+        random.seed(self.seed)
+        tokens_to_mask = random.sample(
+            [
+                token
+                for i, token in enumerate(tokens)
+                if pos_tags[i] in self.pos_to_change
+            ],
+            num_word_swaps,
+        )
+        for token in tokens_to_mask:
+            masked_input = input.replace(token, "<mask>", 1)
+            perturbation_dict[token] = self.unmasker(masked_input)
+
+        return perturbation_dict
+
+    def select_and_apply_perturbations(self, input: str):
+        """
+        Applies perturbations and returns the perturbed sentence
+        By default, the best fitting candidate is used.
+        If the top k candidate list includes a token with the same POS tag but different word family
+        as the original token, it will be used instead of the default.
+        """
+        perturbation_dict = self.get_perturbation_candidates(input)
+        tokens, pos_tags, lemmas = self.get_linguistic_features(input)
+
+        replacement_dict = {}
+        for original_token in perturbation_dict:
+            replacement_dict[original_token] = perturbation_dict[
+                original_token
+            ][0]["token_str"]
+
+            # Replace with the first best choice in case no better candidate is found
```

Comment:
Could you implement returning a list of perturbed sentences with the length of self.max_outputs (If there are not that many outputs you can just return those meeting the POS requirements)?

Reply:
Good point, thank you. I just added the functionality, the transformation now returns the number of permutations specified in self.max_outputs (or alternatively all valid permutations if that number is smaller than self.max_outputs).